code/modules/mob/abstract/unauthed/login.dm

```diff
@@ -0,0 +1,58 @@
+/var/list/unauthed = list()
+
+/mob/abstract/unauthed
+	authed = FALSE
+	var/token = ""
+
+/mob/abstract/unauthed/New()
+	verbs -= typesof(/mob/verb)
+
+/mob/abstract/unauthed/Login()
+	update_Login_details()
+	to_chat(src, "<span class='danger'><b>You need to authenticate before you can continue.</b></span>")
+	token = md5("[client.ckey][client.computer_id][world.time][rand()]")
+	unauthed[token] = src
+	client.verbs -= typesof(/client/verb)
+	var/uihtml = "<html><head><style>body * {display: block;text-align:center;margin: 14px 0;font-size:24px;text-decoration:none;font-family:Segoe UI,Frutiger,Frutiger Linotype,Dejavu Sans,Helvetica Neue,Arial,sans-serif;}</style></head><body><p>Please select:</p>"
+	if(config.guests_allowed)
+		uihtml += "<a href='?src=\ref[src];authaction=guest'>Login as guest</a>"
+	if(config.webint_url)
+		uihtml += "<a href='?src=\ref[src];authaction=forums'>Login via forums</a>"
+	if(!config.guests_allowed && config.webint_url)
+		src.OpenForumAuthWindow()
+	show_browser(src, uihtml, "window=auth;size=300x300;border=0;can_close=0;can_resize=0;can_minimize=0;titlebar=1")
+
+/mob/abstract/unauthed/proc/ClientLogin(var/newkey)
```

Comment:
The presence of this argument here is weird to me.

It might be better to split the functionality a bit? Have these three procs:
* `ClientAuthenticateCkey(newkey)`
* `ClientAuthenticateGuest()`
* `ClientAuthenticateImpl()`

And have the two first ones call the last one for the shared code. Technically you can just make do with 2 tbh:
* `ClientAuthenticateCkey(newkey)`
* `ClientFinishAuthentication()`

wherein the only thing really done by `ClientAuthenticateCkey(newkey)` is the setting of the ckey and invoking `ClientFinishAuthentication`

Reply:
First part need to work with client more and during it, mob may lose it's reference to client
Second part of proc finishes up with mob, during with if put inside client, there may not be a reference to mob any more,

Reply:
I yield.